rotkehlchen/balances/historical.py

```diff
@@ -1,29 +1,182 @@
+from rotkehlchen.globaldb.handler import GlobalDBHandler
+from rotkehlchen.history.events.structures.types import (
+    EventDirection,
+    HistoryEventSubType,
+    HistoryEventType,
+)
 from rotkehlchen.history.price import PriceHistorian
-from rotkehlchen.types import Timestamp
-from rotkehlchen.utils.misc import ts_ms_to_sec
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.serialization.deserialize import deserialize_fval, deserialize_timestamp
+from rotkehlchen.types import Timestamp, TradeType
+from rotkehlchen.utils.misc import timestamp_to_daystart_timestamp
 
 if TYPE_CHECKING:
     from rotkehlchen.db.dbhandler import DBHandler
-    from rotkehlchen.history.events.structures.base import HistoryBaseEntry
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
 
 
 class HistoricalBalance(TypedDict):
     amount: FVal
     price: FVal
 
 
+@dataclass(init=True, repr=True, eq=True, order=False, unsafe_hash=False, frozen=True)
+class MinimalTrade:
+    """Minimal trade event used for balance tracking. Records buys/sells between assets with amount and rate."""  # noqa: E501
+    timestamp: Timestamp
+    identifier: str
+    amount: FVal
+    trade_type: TradeType
+    base_asset: Asset
+    quote_asset: Asset
+    rate: FVal
+
+    @classmethod
+    def deserialize(cls, entry: tuple[Any, ...]) -> 'MinimalTrade':
+        """
+        May raise:
+        - UnknownAsset if the asset in the event cannot be found in the DB.
+        - DeserializationError
+        """
+        return cls(
+            timestamp=deserialize_timestamp(entry[0]),
+            identifier=entry[1],
+            trade_type=TradeType.deserialize_from_db(entry[2]),
+            base_asset=Asset(entry[3]).check_existence(),
+            quote_asset=Asset(entry[4]).check_existence(),
+            amount=deserialize_fval(
+                value=entry[5],
+                name='amount',
+                location='historical balances',
+            ),
+            rate=deserialize_fval(
+                value=entry[6],
+                name='rate',
+                location='historical balances',
+            ),
+        )
+
+    def update_balances(self, current_balances: dict[Asset, FVal]) -> str | None:
+        """Updates current balances for a trade event, checking for negative balances.
+
+        Returns the event_identifier if a negative balance would occur, else None.
+        For trades:
+        - BUY means we receive base_asset and spend quote_asset
+        - SELL means we spend base_asset and receive quote_asset
+        """
+        is_buy = self.trade_type in {TradeType.BUY, TradeType.SETTLEMENT_BUY}
+        spend_asset = self.quote_asset if is_buy else self.base_asset
+        receive_asset = self.base_asset if is_buy else self.quote_asset
+        spend_amount = self.amount * self.rate if is_buy else self.amount
+        receive_amount = self.amount if is_buy else self.amount * self.rate
```

Comment:
donyou can use a normal if/else. Otherwise you are performing the same check 4 times